### PR TITLE
Add full export history page (with pagination)

### PR DESCRIPTION
**Pull request by debitan**
## Description of change

This adds pagination to the full export history page (previous PR - https://github.com/uktrade/data-hub-frontend/pull/2420). Please only focus on the `add pagination and axios request` commit in this PR. The other commits will disappear once I merged my previous PRs.

This is a hidden URL at the moment. I will create a seperate PR to add the link to this page to the export page. 

All additions and deletions are currently shown as individual events. There is a ticket for the backend to group these together to reflect when countries were added and deleted in a single user event. I will further iterate this front end feature to incorporate this when it is ready. 

## Test instructions

You should see the export countries history using the collectionList component when going to /companies/:companyId/exports/history

There is only history from when we merged the history PR last week - so you may have to add and delete some export countries to see the feature in action. 

### Before 

FireShot Capture 034 - Full export history - Exports - One List Subsidiary Ltd - Companies -_ - localhost

#### Page 1
FireShot Capture 035 - Full export history - Exports - One List Subsidiary Ltd - Companies -_ - localhost

#### Page 2
FireShot Capture 036 - Full export history - Exports - One List Subsidiary Ltd - Companies -_ - localhost

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"

- [ ] Has the branch been rebased to master?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: test/functional/cypress/specs/companies/export-spec.js [hunk 1]
```diff
@@ -202,8 +203,146 @@ describe('Companies Export', () => {
...
+      cy.contains('12 results').should('exist')
+      cy.get(countrySelectors.listItemHeadings).should('have.length', 10)
+
+      cy.contains('Belarus added to countries of no interest').should('exist')
+      cy.contains('Belarus added to countries of no interest')
+        .siblings()
+        .should('contain', 'By DIT Staff')
+        .should('contain', 'Date 11 Feb 2020, 10:47am')
+      cy.contains('Botswana added to future countries of interest').should(
+        'exist'
```
Comment (peterhudec) [inline test/functional/cypress/specs/companies/export-spec.js]: No need for `.should('exist')`

File: test/functional/cypress/specs/companies/export-spec.js [hunk 2]
```diff
@@ -202,8 +203,146 @@ describe('Companies Export', () => {
...
+      )
+      cy.contains('Botswana added to future countries of interest')
+        .siblings()
+        .should('contain', 'By DIT Staff')
+        .should('contain', 'Date 11 Feb 2020, 10:47am')
+      cy.contains('Georgia added to currently exporting').should('exist')
```
Comment (peterhudec) [inline test/functional/cypress/specs/companies/export-spec.js]: No need for `.should('exist')`, till the end of the file...

File: src/apps/companies/apps/exports/client/ExportsFullHistory.jsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  useEffect(() => {
+    async function fetchData() {
+      try {
+        const { data } = await axios.get(urls.search.exportsHistory(), {
+          params: {
+            page: activePage,
+            companyId,
+          },
+        })
+        setitems(data.results)
```
Comment (rafenden) [inline src/apps/companies/apps/exports/client/ExportsFullHistory.jsx]: Not a big deal but the function should have been named `setItems` (capital "I")

File: src/apps/companies/apps/exports/views/full-history.njk [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{% extends "../../../views/template.njk" %}
+
+{% from "details/macro.njk" import govukDetails %}
+
+{% block body_main_content %}
+
+  <div class="govuk-grid-column-full">
+            {% component 'react-slot', {
```
Comment (rafenden) [inline src/apps/companies/apps/exports/views/full-history.njk]: The code formatting is a bit broken here.

File: test/functional/cypress/specs/companies/export-spec.js [hunk 5]
```diff
@@ -202,8 +203,146 @@ describe('Companies Export', () => {
...
+
+      cy.contains('Belarus added to countries of no interest').should('exist')
+      cy.contains('Belarus added to countries of no interest')
+        .siblings()
+        .should('contain', 'By DIT Staff')
+        .should('contain', 'Date 11 Feb 2020, 10:47am')
+      cy.contains('Botswana added to future countries of interest').should(
+        'exist'
+      )
+      cy.contains('Botswana added to future countries of interest')
```
Comment (rafenden) [inline test/functional/cypress/specs/companies/export-spec.js]: Perhaps, it's worth creating a helper function as it will be useful when testing collections in other places? e.g.
```
assertCollection([
  {
    headingText: 'Georgia added to currently exporting', metadata: [
      { 'By': 'DIT Staff' },
      { 'Date': '11 Feb 2020, 10:47am' }
    ],
  },
  {},
])
```

File: src/apps/companies/apps/exports/controllers.js [hunk 6]
```diff
@@ -86,17 +82,15 @@ function renderExports(req, res) {
 }
 
 async function renderFullExportHistory(req, res) {
-  const { token } = req.session
   const { company } = res.locals
```
Comment (ian-leggett) [inline src/apps/companies/apps/exports/controllers.js]: Optional - You could do `const {company: {name, id}} = res.locals` then you can use the two variables name and id.

**Comment by debitan** [comment]
> When I try and get the export history I get an error message, should this just be a page with no results?
> Screenshot 2020-02-18 at 09 26 31

we discussed. this was from the old deployment. new deployment should work soon as I've now rebased on the changes for postgres docker image.

File: src/apps/companies/apps/exports/tasks.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+function getCountryText(country, historyType, status) {
+  const historyTypeText = COUNTRY_HISTORY_TYPE_TEXT[historyType]
+  const typeText = COUNTRY_TYPE_TEXT[status]
+
+  return [country, historyTypeText, typeText].join(' ')
+}
+
+function createCountry(item) {
```
Comment (ian-leggett) [inline src/apps/companies/apps/exports/tasks.js]: You commit use an implicit return here, could we use 
`
const createCountry = (item)=> ({
    headingText: getCountryText(
      item.country.name,
      item.history_type,
      item.status
    ),
    metadata: [
      { label: 'By', value: item.history_user.name },
      { label: 'Date', value: formatDateTime(item.history_date) },
    ],
  })
`

File: src/server.js [hunk 8]
```diff
@@ -140,12 +137,23 @@ app.use(
       pathRewrite: {
         ['^' + API_PROXY_PATH]: '',
       },
-      onProxyReq: (proxyReq, req) =>
-        proxyReq.setHeader('authorization', `Bearer ${req.session.token}`),
+      onProxyReq: (proxyReq, req) => {
```
Comment (ian-leggett) [inline src/server.js]: Does this need to be in this PR?

File: test/sandbox/routes/v4/search/export.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+var fullExportHistoryPage1 = require('../../../fixtures/v4/export/fullExportHistoryPage1.json')
+var fullExportHistoryPage2 = require('../../../fixtures/v4/export/fullExportHistoryPage2.json')
+var emptyFullExportHistory = require('../../../fixtures/v4/export/emptyFullExportHistory.json')
+
+exports.fetchFullExportHistory = function(req, res) {
+  if (
+    req.body.company === 'cc7e2f19-7251-4a41-a27a-f98437720531' &&
```
Comment (ian-leggett) [inline test/sandbox/routes/v4/search/export.js]: Not a big one but should this id reference `dnbCorp.id` from fixtures?

File: src/server.js (hunk 8 above)
Comment (debitan) [reply src/server.js]: I can split it out if required. It is needed in the sense that the feature doesn't work without this change to the proxy.

File: src/apps/companies/apps/exports/tasks.js (hunk 7 above)
Comment (web-bert) [reply src/apps/companies/apps/exports/tasks.js]: I prefer to see functions as functions

Comment (debitan) [reply src/apps/companies/apps/exports/tasks.js]: > I have tested and it works, although I would rather it didn't re-render the whole component list when changing pages - the heading and the results count do not need to be re-rendered
> 
> Also the URL url is not being updated when changing pages

@web-bert I've added back in the function to update the url on page changes. Regarding re-rendering the whole component - I think this is a consequence of the task component fetching the data. I'm not sure I can do much about that in the scope of this PR. Maybe something to look at together @peterhudec @ian-leggett ?

Comment (peterhudec) [reply src/apps/companies/apps/exports/tasks.js]: @web-bert Any UX pattern can be accomodated, but for the _translucent_ loader to pattern we would need to create a new component similar to `Task.Status`. The question is only whether it is in scope of this PR?

File: src/server.js (hunk 8 above)
Comment (peterhudec) [reply src/server.js]: It only ads about 13 lines of code to this PR. Separating it would create a procedural overhead.

File: src/server.js [hunk 10]
```diff
@@ -140,12 +137,23 @@ app.use(
...
+        proxyReq.setHeader('authorization', `Bearer ${req.session.token}`)
+        if (req.body) {
```
Comment (peterhudec) [inline src/server.js]: I would perhaps add a little explanation comment about why is this necessary with a URL of the source of the solution

File: src/apps/companies/apps/exports/client/ExportsFullHistory.jsx [hunk 11]
```diff
@@ -1,29 +1,67 @@
...
+        type: EXPORTS_HISTORY__CLICK,
+        page,
+        event,
+      })
+      window.history.pushState(
+        {},
+        '',
+        `${window.location.pathname}?page=${page}`
+      )
+      window.scrollTo(0, 0)
```
Comment (peterhudec) [inline src/apps/companies/apps/exports/client/ExportsFullHistory.jsx]: These are side effects and should be done in a saga

File: src/apps/companies/apps/exports/tasks.js (hunk 7 above)
Comment (debitan) [reply src/apps/companies/apps/exports/tasks.js]: I will update them to function expressions just to keep the file consistent.

File: src/apps/companies/apps/exports/client/ExportsFullHistory.jsx [hunk 12]
```diff
@@ -1,29 +1,62 @@
...
 const Wrapper = styled('div')`
   margin-top: ${SPACING.SCALE_3};
 `
 
-function ExportFullHistory({ data }) {
+export default connect(
+  (state) => {
+    const { loading, count, results, activePage } = state2props(state)
+    return { loading, count, results, activePage }
+  },
```
[the remark this answers is not in the document]
Comment (debitan) [reply src/apps/companies/apps/exports/client/ExportsFullHistory.jsx]: Yeah, this was a linter thing. Just did it for me.

[the remark this answers is not in the document]
Comment (debitan) [reply src/apps/companies/apps/exports/client/ExportsFullHistory.jsx]: so might just leave it. it does seem more readable this way.

File: src/apps/companies/apps/exports/client/ExportsFullHistory.jsx [hunk 13]
```diff
@@ -1,29 +1,62 @@
...
+      <Task.Status
+        name="Exports history"
+        id="exportsHistory"
+        progressMessage="loading Exports History"
+        startOnRender={{
+          payload: { companyId, activePage },
+          onSuccessDispatch: EXPORTS_HISTORY,
+        }}
       />
+      {!loading && (
```
[the remark this answers is not in the document]
Comment (debitan) [reply src/apps/companies/apps/exports/client/ExportsFullHistory.jsx]: that example doesn't seem to work. I would also say that have `!loading` there makes it clearer?

File: src/apps/companies/apps/exports/reducer.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+export default (state = initialState, { type, result, page }) => {
+  switch (type) {
+    case EXPORTS_HISTORY:
+      const { count, results } = result
+      return {
+        ...state,
+        loading: false,
+        count,
+        results,
```
[the remark this answers is not in the document]
Comment (debitan) [reply src/apps/companies/apps/exports/reducer.js]: to be contrary, this is exactly what it is called in the api though. there is an argument for keeping consistency.

**Comment by web-bert** [comment]
@peterhudec the comments have got all messed up, so replying here instead. The way it is now is really janky, the loading status should not be coming in above the component but below the results - as this causes the content to shift about unnecessarily. As this is hidden at the moment it can be fixed in a different PR, but I think it needs to be fixed as a pattern...

export_history

**Comment by web-bert** [comment]
@debitan it now updates the URL but when I go back and forwards it does not render the changes...
